### PR TITLE
Issue #367: Solved column order of struct bug. Now the column order of struct variables need not be same as that of BQ schema

### PR DESCRIPTION
**Pull request by himanshukohli09**

### CODE REVIEW

File: connector/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java [hunk 1]
```diff
@@ -61,6 +69,14 @@ public ArrowBinaryIterator(
...
     currentIterator = ImmutableList.<InternalRow>of().iterator();
     this.columnsInOrder = columnsInOrder;
+
+    List<StructField> userProvidedFieldList =
+        Arrays
+            .stream(userProvidedSchema.orElse(new StructType()).fields())
+            .collect(Collectors.toList());
+
+    this.userProvidedFieldMap =
+        userProvidedFieldList.stream().collect(Collectors.toMap(StructField::name, field -> field));
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/ArrowBinaryIterator.java:79]: done

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java [hunk 2]
```diff
@@ -113,53 +121,82 @@ static Object convert(Field field, Object value) {
...
+          new StructField(
+              field.name(),
+              arrayType.elementType(),
+              arrayType.containsNull(),
+              Metadata.empty()); // safe to pass empty metadata as it is not used anywhere
+    }
+    return nestedField;
+  }
+
+  static Object convertByBigQueryType(Field bqField, Object value, StructField userProvidedField) {
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java:152]: renamed it so as to have a clearer distinction between the two fields used, one from BQ and one provided by user. Earlier this method did not have the userProvidedField so no such distinction was required.

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java [hunk 3]
```diff
@@ -113,53 +121,82 @@ static Object convert(Field field, Object value) {
...
+      List<String> namesInOrder = null;
+      List<StructField> structList = null;
+
+      if (userProvidedField != null) {
+        structList =
+            Arrays
+                .stream(((StructType)userProvidedField.dataType()).fields())
+                .collect(Collectors.toList());
+
+        namesInOrder = structList.stream().map(f -> f.name()).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java:190]: done

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java [hunk 4]
```diff
@@ -113,53 +121,82 @@ static Object convert(Field field, Object value) {
...
+      } else {
+        namesInOrder =
+            bqField.getSubFields().stream().map(f -> f.getName()).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java:193]: done

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java [hunk 5]
```diff
@@ -171,13 +208,28 @@ static Object convertByBigQueryType(Field field, Object value) {
...
+      List<String> namesInOrder,
+      List<StructField> userProvidedFieldList) {
     Map<String, Object> fieldMap = new HashMap<>();
 
+    Map<String, StructField> userProvidedFieldMap =
+        userProvidedFieldList == null
+            ? new HashMap<>()
+            : userProvidedFieldList
+                .stream()
+                .collect(Collectors.toMap(StructField::name, field -> field));
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java:222]: done

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**